compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -631,6 +632,13 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         goal: Goal<'tcx, G>,
         candidates: &mut Vec<Candidate<'tcx>>,
     ) {
+        if self.tcx().has_attr(
+            goal.predicate.trait_def_id(self.tcx()),
+            sym::rustc_do_not_implement_via_object,
+        ) {
```

Comment:
please instead add another field to `TraitDef` and access that here. you can then mark `rustc_do_not_implement_via_object` as `@only_local`.